### PR TITLE
build(bazel): add a bazel aio serve target

### PR DESCRIPTION
**Pull request by kormide**
@josephperrott @devversion @gkalpak @gregmagolan 

This works with ibazel but is not highly incremental

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the new behavior?

Add a bazel aio serve target that works with ibazel. It's not very incremental as it has to re-do the full dgeni build each time something changes. Still thinking about how to make that more incremental---more thoughts to come on that later.

Can serve the app locally with `ibazel run //aio:serve`.

Includes all doc content except for stackblitz links and example zips (to come later).

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -10,8 +10,8 @@ load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
 
 http_archive(
     name = "build_bazel_rules_nodejs",
-    sha256 = "3ceb1e5b5dcad5fa2ad8870a20201cfbb9c9c63cac4055c9ab370034c765297f",
-    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/5.3.0/rules_nodejs-5.3.0.tar.gz"],
+    sha256 = "523da2d6b50bc00eaf14b00ed28b1a366b3ab456e14131e9812558b26599125c",
+    urls = ["https://github.com/bazelbuild/rules_nodejs/releases/download/5.3.1/rules_nodejs-5.3.1.tar.gz"],
```
Comment (kormide) [inline WORKSPACE:14]: Needed a fix for [chdir](https://github.com/bazelbuild/rules_nodejs/commit/9ed16c076e5ea10237873be592d5213e190683b8) to work correctly.

File: aio/BUILD.bazel [hunk 2]
```diff
@@ -139,11 +140,11 @@ DGENI_DEPS = [
 
 # All transforms and sources needed by dgeni to produce docs
 DGENI_FILES = glob([
-    "content/**",
     "src/assets/**",
     "tools/transforms/**",
```
Comment (kormide) [inline aio/BUILD.bazel:144]: If I'm including everything in these folders I suppose I should add targets for them too. Or would you like me to more carefully declare what gets included? I'll double check that we actually need _everything_..

File: aio/BUILD.bazel [hunk 3]
```diff
@@ -138,12 +141,11 @@ DGENI_DEPS = [
...
-DGENI_FILES = glob([
-    "content/**",
-    "src/assets/**",
-    "tools/transforms/**",
-]) + [
+DGENI_FILES = [
     "@angular_cli_src//:files_for_docgen",  # Downloaded cli snapshot (see WORKSPACE)
+    "//aio/content",
+    "//aio/tools/transforms",
+    "//aio/src/assets",
```
Comment (gkalpak) [inline aio/BUILD.bazel:148]: Nit: 😇

```suggestion
    "//aio/src/assets",
    "//aio/tools/transforms",
```

File: aio/angular.json [hunk 4]
```diff
@@ -51,7 +51,7 @@
               "src/pwa-manifest.json",
               "src/google385281288605d160.html",
               {
-                "input": "src/generated",
+                "input": "dgeni/generated",
```
Comment (gkalpak) [inline aio/angular.json:54]: Why is this change needed?

File: aio/BUILD.bazel [hunk 5]
```diff
@@ -160,3 +162,16 @@ npm_package_bin(
...
+    args = [
+        "site:serve",
+    ],
+    chdir = package_name(),
+    data = APPLICATION_DEPS + APPLICATION_FILES,
+    env = {
+        "BAZEL_DGENI_OUTPUT_PATH": "$(@D)",
+    },
+    tags = ["ibazel_notify_changes"],
+)
```
Comment (gkalpak) [inline aio/BUILD.bazel:177]: Could we somehow leverage [docs-watch][1] here?

[1]: https://github.com/angular/angular/blob/035ba83b16772b7322745f9a5308b3287f5726bc/aio/package.json#L56

File: aio/angular.json (hunk 4 above)
Comment (kormide) [reply aio/angular.json:54]: This is needed because the dgeni docs are no longer under the source tree. The docs get built and placed in the bazel output dir, and are included as runfiles in this architect serve rule, so they are accessible under this `dgeni` dir.

File: aio/BUILD.bazel (hunk 5 above)
Comment (kormide) [reply aio/BUILD.bazel:174]: I plan to look at incrementality after getting the stackblitz and example zips working. Chances are it will need to work a bit differently under bazel, but I want to think about it some more and present some options to you guys.

File: aio/BUILD.bazel [hunk 6]
```diff
@@ -34,6 +34,9 @@ APPLICATION_FILES = [
     "tsconfig.app.json",
     "tsconfig.json",
     "tsconfig.worker.json",
+    "//aio/src/assets",
+    "//aio/src/assets/js",
```
Comment (gkalpak) [inline aio/BUILD.bazel:38]: Why don't we make `assets/js` part of `//aio/src/assets` (as we do for `assets/images`, for example)?

File: aio/angular.json (hunk 4 above)
Comment (kormide) [reply aio/angular.json:54]: You mean the live examples and example zips? Those are coming up in the next pr, and will have their own entry in `angular.json`.

File: aio/BUILD.bazel (hunk 6 above)
Comment (kormide) [reply aio/BUILD.bazel:38]: I split those up because `assets` are needed by both the app and dgeni generation, whereas `assets/js` is only needed by the app. So `assets/js` changes, bazel won't try to rebuild dgeni.

File: aio/angular.json (hunk 4 above)
Comment (gkalpak) [reply aio/angular.json:54]: I see :+1:
Yes, currently it is live examples and zips, but there could be others in the future (for example, I am adding another file in a PR I have in progress).

Not necessarily for this one, but in general I think we are at a point where it would make sense to try to have each PR in a (mostly) working state, because otherwise we will end up with a completely broken setup and it would be much more difficult to get it working.

Comment (gkalpak) [reply aio/angular.json:54]: OK, you know better. Let's move forward with the plan you have in mind. I was just thinking out loud :smiley:

**Review by gkalpak — APPROVED** [review]

File: aio/content/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,9 @@
+package(default_visibility = ["//visibility:public"])
+
+filegroup(
+    name = "content",
+    srcs = glob(
+        ["**"],
+        exclude = ["**/BUILD.bazel"],
```
Comment (devversion) [inline aio/content/BUILD.bazel:7]: Would it ever match such build files? The glob would not pierce into the nested package anyway right? (or are some files bazel-ignored? -- I see that the current package is ignored, likely should be removed now)

File: aio/tools/transforms/BUILD.bazel [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+        [
+            "**/*.js",
+            "**/*.html",
+            "**/*.json",
+            "**/*.xml",
+        ],
+        exclude = [
+            "test.js",
+            "**/*.spec.js",
+            "authors-package",
```
Comment (devversion) [inline aio/tools/transforms/BUILD.bazel:15]: Is there any reason why the authors package is excluded? if so, please add a comment

**Review by devversion — APPROVED** [review]
LGTM, just two minor comments

File: aio/content/BUILD.bazel (hunk 7 above)
Comment (kormide) [reply aio/content/BUILD.bazel:7]: It globs the BUILD file in the root of content. But you're right that it would never match anything lower because they would be their own package, so I'll change this to just exclude the root build file.

File: aio/tools/transforms/BUILD.bazel (hunk 8 above)
Comment (kormide) [reply aio/tools/transforms/BUILD.bazel:15]: The authors package isn't used for the high-fidelity dgeni generation. It's a set of ad-hoc packages that are used by the docs-watcher for fast builds after the main build has completed.

Comment (kormide) [reply aio/tools/transforms/BUILD.bazel:16]: comment added

**Review by devversion — APPROVED** [review]

**Review by josephperrott — APPROVED** [review]
LGTM
